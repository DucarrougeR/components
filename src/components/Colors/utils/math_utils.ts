/**
 * Utility to convert a numerical value to a percent given the value and a max value.
 */
export const toPerc = (value: number, maxValue: number) =>
  Math.round((value / maxValue) * 100)
