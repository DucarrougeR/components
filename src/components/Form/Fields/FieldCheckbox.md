
```js noeditor
const StatusAndResources = require('../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/Fields/FieldCheckbox.tsx"
  feedbackTitle="FieldCheckbox Component Feedback" />
```


```js
  <FieldCheckbox id="theme-name" name="themeName" label="Theme Name" />
  <FieldCheckbox id="theme-name2" name="themeName2" label="Theme Name 2" />
  <FieldCheckbox id="theme-name3" name="themeName3" label="Theme Name 3" />
```
