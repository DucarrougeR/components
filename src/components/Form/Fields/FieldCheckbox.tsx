import * as React from 'react'
import { FormControl, FormControlDirections } from '../FormGroup/FormGroup'
import { InputCheckbox } from '../Inputs/InputCheckbox'
import { Label } from '../Label/Label'

export interface FieldCheckboxProps {
  id: string
  label: string
  name: string
}

export const FieldCheckbox: React.SFC<FieldCheckboxProps> = ({
  id,
  label,
  name,
  ...props
}) => {
  return (
    <FormControl alignLabel={FormControlDirections.Left} alignCenter>
      <Label htmlFor={id}>{label}</Label>
      <InputCheckbox id={id} name={name} {...props} />
    </FormControl>
  )
}
