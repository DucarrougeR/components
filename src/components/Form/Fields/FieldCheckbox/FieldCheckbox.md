
```js noeditor
const StatusAndResources = require('../../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/Fields/FieldCheckbox.tsx"
  feedbackTitle="FieldCheckbox Component Feedback" />
```

### FieldCheckbox

The `<FieldCheckbox />` component has one required property: `label`. By default, the label will render to the left of the checkbox, which is the recommended convention, however this is adjustable with the `alignLabel` property.

Additionally, it is highly recommended to include the `name` property, as this will be used by forms to capture the inputted data and for validation.

The `required` property displays a red star by the label, indicating that the field must be filled out before submission.

```js
<FieldCheckbox name="box" label="Box" />
<FieldCheckbox name="box-2" label="Box 2" alignLabel="left" />
<FieldCheckbox name="requiredField" label="A required field" required />
```

### FieldCheckbox Validation

The `<FieldCheckbox />` components accepts `validationMessage` as a property. This is an object with properties `type` and `message`:
`
interface ValidationMessage {
  type: ValidationType
  message: string
}
`

`ValidationType` is a string literal, which accepts:
* error
* warning
* info
* success


If you have multiple fields in need of validation, it is recommended that you wrap them in a `<Form />` component and pass to it a dictionary with key as field name and value as a `ValidationMessage`.

The `alignValidationMessage` can also be used to adjust where the validation message appears in relation to the input field. By default, it will render to the right, which is the best practice in most cases.

___Note: In the current implementation, any `ValidationMessage` set in a `<FieldCheckbox />` component will be overridden by `ValidationMessages` set in the parent `<Form />` component. Because of this, it is highly recommended that you only ever directly set `ValidationMessage` if a `<FieldCheckbox />` is outside a `<Form />`.___

```js
<FieldCheckbox name="someBox" label="Some Box" validationMessage={{type: "error", message: "This is an error"}}/>
<FieldCheckbox name="some-box-2" label="Some Box 2" validationMessage={{type: "warning", message: "Here is a warning"}} alignValidationMessage="bottom" />
```
