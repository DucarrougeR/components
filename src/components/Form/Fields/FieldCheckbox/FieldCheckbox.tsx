import * as React from 'react'
import uuid from 'uuid/v4'
import { styled } from '../../../../style'
import { withForm } from '../../Form'
import { Checkbox, CheckboxProps } from '../../Inputs/Checkbox/Checkbox'
import { Field, FieldProps } from '../Field'

interface FieldCheckboxProps extends FieldProps, CheckboxProps {}

const InternalFieldCheckbox = (props: FieldCheckboxProps) => {
  const {
    alignLabel,
    alignValidationMessage,
    label,
    labelWidth,
    labelFontWeight,
    validationMessage,
    id = uuid(),
    ...checkboxProps
  } = props
  return (
    <Field
      {...props}
      id={id}
      alignLabel={alignLabel ? alignLabel : 'left'}
      alignValidationMessage={
        alignValidationMessage ? alignValidationMessage : 'right'
      }
    >
      <Checkbox
        id={id}
        {...checkboxProps}
        validationType={validationMessage && validationMessage.type}
      />
    </Field>
  )
}

const FieldCheckboxFactory = React.forwardRef((props: FieldProps, ref) => (
  <InternalFieldCheckbox
    innerRef={ref as React.RefObject<HTMLElement>}
    {...props}
  />
))

export const FieldCheckbox = styled(withForm(FieldCheckboxFactory))``
