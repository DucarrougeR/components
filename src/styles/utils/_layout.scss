// Returns a value from the $spacing list
@function spacing($key) {
  @return map-lookup($spacing, $key, '$spacing');
}
