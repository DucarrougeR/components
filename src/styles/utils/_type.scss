// Returns a value from the $test-size list
@function font-size($key) {
  @return map-lookup($text-size, $key, '$text-size');
 }

 // Returns a value from the $line-height list
 @function line-height($key) {
   @return map-lookup($line-height, $key, '$line-height');
 }

 @mixin fontSize($size) {
  font-size: $size;
  font-size: pxToRem($size);
}
