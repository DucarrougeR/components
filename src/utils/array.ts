/**
 * Returns an array of length `size` filled with zeros.
 * Useful for looping constructs where a length is known, but no array is provided.
 */

export const sizedArray = (size: number) => Array(size).fill(0)
