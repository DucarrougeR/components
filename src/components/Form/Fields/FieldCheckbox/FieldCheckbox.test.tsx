import 'jest-styled-components'
import * as React from 'react'
import { createWithTheme } from '../../../../../test/utils/create_with_theme'
import { FieldCheckbox } from './FieldCheckbox'

test('A FieldCheckbox', () => {
  const component = createWithTheme(
    <FieldCheckbox label="👍" name="thumbsUp" id="thumbs-up" />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A required FieldCheckbox', () => {
  const component = createWithTheme(
    <FieldCheckbox label="👍" name="thumbsUp" id="thumbs-up" required />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A FieldCheckbox with an error validation aligned to the bottom', () => {
  const component = createWithTheme(
    <FieldCheckbox
      label="👍"
      name="thumbsUp"
      id="thumbs-up"
      validationMessage={{ type: 'error', message: 'This is an error' }}
      alignValidationMessage="bottom"
    />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(tree!.children![0].props.htmlFor).toEqual(
    tree!.children![1].children![1].props.id
  )
})
