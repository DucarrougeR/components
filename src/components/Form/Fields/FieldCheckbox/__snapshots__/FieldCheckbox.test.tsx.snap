// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A FieldCheckbox 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 0.5rem;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  border: 0;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-weight: 600;
  color: #262D33;
  margin-right: 0.5rem;
  font-size: 0.875rem;
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="thumbs-up"
  >
    üëç
  </label>
  <div
    className="c2"
  >
    <div
      className=""
    >
      <input
        id="thumbs-up"
        name="thumbsUp"
        type="checkbox"
      />
    </div>
  </div>
</div>
`;

exports[`A FieldCheckbox with an error validation aligned to the bottom 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 0.5rem;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  border: 0;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-weight: 600;
  color: #262D33;
  margin-right: 0.5rem;
  font-size: 0.875rem;
}

.c4 {
  margin-right: 0.5rem;
  margin-top: 0.5rem;
  font-size: 0.75rem;
  color: #B2121F;
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="thumbs-up"
  >
    üëç
  </label>
  <div
    className="c2 c3"
  >
    <div
      className=""
    >
      <input
        id="thumbs-up"
        name="thumbsUp"
        type="checkbox"
      />
    </div>
    <div
      className=""
    >
      <div
        className="c4"
      >
        This is an error
      </div>
    </div>
  </div>
</div>
`;

exports[`A required FieldCheckbox 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 0.5rem;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  border: 0;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  font-weight: 600;
  color: #262D33;
  margin-right: 0.5rem;
  font-size: 0.875rem;
}

.c2 {
  color: #990F14;
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="thumbs-up"
  >
    üëç
    <span
      className="c2"
    >
       *
    </span>
  </label>
  <div
    className="c3"
  >
    <div
      className=""
    >
      <input
        id="thumbs-up"
        name="thumbsUp"
        required={true}
        type="checkbox"
      />
    </div>
  </div>
</div>
`;
